### ISSUE TITLE
[ARROY] Corruption: panic using hybrid search

### ISSUE THREAD

**Issue by sam-ulrich1** [meilisearch/meilisearch#4588]
I've confirmed that there is a valid embedding being sent and a valid query

```
2024-04-18T22:38:19.654577Z ERROR meilisearch: info=panicked at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/arroy-0.2.0/src/reader.rs:250:83: called `Option::unwrap()` on a `None` value
2024-04-18T22:38:19.654906Z ERROR HTTP request{method=POST host="meili-meilisearch.gigo-core.svc.cluster.local:7700" route=/indexes/bytes/search query_parameters= user_agent=Meilisearch Go (v0.26.2) status_code=500 error=task 595 panicked}: tracing_actix_web::middleware: Error encountered while processing the incoming HTTP request: ResponseError { code: 500, message: "task 595 panicked", error_code: "internal", error_type: "internal", error_link: "https://docs.meilisearch.com/errors#internal" }
2024-04-18T22:38:19.655012294Z 2024-04-18T22:38:19.654962Z  INFO HTTP request{method=POST host="meili-meilisearch.gigo-core.svc.cluster.local:7700" route=/indexes/bytes/search query_parameters= user_agent=Meilisearch Go (v0.26.2) status_code=500 error=task 595 panicked}: meilisearch: close time.busy=669µs time.idle=7.79ms
2024-04-18T22:38:19.658533Z ERROR meilisearch: info=panicked at /usr/local/cargo/registry/src/index.crates.io-6f17d22bba15001f/arroy-0.2.0/src/reader.rs:250:83: called `Option::unwrap()` on a `None` value
2024-04-18T22:38:19.658727629Z 2024-04-18T22:38:19.658687Z ERROR HTTP request{method=POST host="meili-meilisearch.gigo-core.svc.cluster.local:7700" route=/indexes/bytes/search query_parameters= user_agent=Meilisearch Go (v0.26.2) status_code=500 error=task 597 panicked}: tracing_actix_web::middleware: Error encountered while processing the incoming HTTP request: ResponseError { code: 500, message: "task 597 panicked", error_code: "internal", error_type: "internal", error_link: "https://docs.meilisearch.com/errors#internal" }
```

**Comment by sam-ulrich1** [comment]
I dropped the whole index and re-created it and now it doesn't happens so it looks like some form of data corruption issue

**Comment by sam-ulrich1** [comment]
@Kerollmops I guess you're confused? Not sure how to take an emoji

**Comment by Kerollmops** [comment]
I am just reacting to the issue. Yes, there is a panic in the engine, it seems related to Arroy and we will fix it soon.

**Comment by irevoire** [comment]
Hey @sam-ulrich1, sorry I’m a little late to the party, but I was wondering, do you have any snapshots or dumps of the data that corrupted your database?

Also, which version of meilisearch were you using, and did you do any updates while you had vectors in your DB?

**Comment by sam-ulrich1** [comment]
I do not have any dumps for this point in time. I can describe the procedure that led to it. Pretty simple.
- New Index
- Ingest ~1000 new vectors @ 1024 dim
- Perform about 30 searches on it

I'm sure something else contributed though because I dropped the index and recreated with the same steps and it did not corrupt again

**Comment by irevoire** [comment]
Thanks a lot for the answer!
Two last questions:
- Which version of meilisearch were you using?
- Did you send the 1000 vectors in one document addition or in multiple additions (and if it’s in multiple additions, do you have an idea of the number of vectors per document addition?)

**Comment by curquiza** [comment]
Closing due to a lack of answer
Feel free to re-open if needed (and I recommend testing with the latest version of Meilisearch 😄)

**Comment by curquiza** [comment]
I reopen because the Db corruption in array is still an issue

**Comment by dureuill** [comment]
I discussed with @irevoire who's working on this yesterday, and we understood why he was not reproducing, and I was.

To reproduce, one must:

1. Add a document with an embedding to the DB
2. Update the document to remove the embedding.
3. Perform a search

Here are the steps performed in arroy at each step:
```
1. adding document
1. 1. attempting to remove item 0 from Writer 0 -> nothing removed, break the loop
1. 2. adding item 0 with vector [0.0, 0.0, 0.0] to Writer 0
1. 3. rebuilding arroy tree for writer 0 as the db is not empty
2. deleting embedding
2. 1. attempting to remove item 0 from Writer 0 -> removed, continue the loop
2. 2. attempting to remove item 0 from Writer 1 -> nothing removed, break the loop
2. 3. NOT rebuilding arroy tree for writer 0 as the db is empty
3. doing a search
```
The issue occurs in step (2. 3.):
- Because the DB was modified by being emptied, the tree needs rebuilding
- But Meilisearch uses the emptiness of the DB to determine if the tree needs rebuilding

Next steps (discussed with @irevoire)

1. Add a `needs_build` method to arroy
2. Use `needs_build` instead of `is_empty` to determine if a tree needs building in Meilisearch
3. (optional) replace the panic with a dedicated error if a tree that needed building wasn't built.

**Comment by curquiza** [comment]
Closed by #4715
